R/pecs.R

```diff
@@ -0,0 +1,139 @@
+#' @title Prediction Error Curves for PredictionSurv and LearnerSurv
+#' @description Methods to plot prediction error curves (pecs) for either a [PredictionSurv] object
+#' or a list of trained [LearnerSurv]s.
+#' @param x ([PredictionSurv] or `list` of [LearnerSurv]s)
+#' @param meas (`character(1)`) \cr
+#'   Either `"graf"` for [MeasureSurvGraf], or `"logloss"` for [MeasureSurvIntLogloss]
+#' @param times (`numeric()`) \cr
+#'   If provided then either a vector of time-points to evaluate `meas` or a range of time-points.
+#' @param n (`integer()`) \cr
+#'   If `times` is missing or given as a range, then `n` provide number of time-points to evaluate
+#'    `meas` over.
+#' @param eps (`numeric()`) \cr
+#'   Small error value to pass to [MeasureSurvIntLogloss] to prevent errors resulting from a log(0)
+#'   calculation.
+#' @param task ([TaskSurv])
+#' @param ... Additional arguments.
+#'
+#' @details If `times` and `n` are missing then `meas` is evaluated over all observed time-points
+#' from the [PredictionSurv] or [TaskSurv] object. If a range is provided for `times` without `n`,
+#' then all time-points between the range are returned.
+#'
+#' @examples
+#' library(mlr3)
+#' task = tsk("rats")
+#'
+#' # Prediction Error Curves for prediction object
+#' learn = lrn("surv.coxph")
+#' p = learn$train(task)$predict(task)
+#' pecs(p)
+#' pecs(p, meas = "logloss", times = c(20, 40, 60, 80)) +
+#'   ggplot2::geom_point() +
+#'   ggplot2::ggtitle("Logloss Prediction Error Curve for Cox PH")
+#'
+#' # Prediction Error Curves for fitted learners
+#' learns = lrns(c("surv.kaplan", "surv.coxph", "surv.ranger"))
+#' lapply(learns, function(x) x$train(task))
+#' pecs(learns, task = task, meas = "logloss", times = c(20, 90), n = 10)
+#'
+#'
+#' @export
+pecs = function(x, meas = c("graf","logloss"), times, n, eps = 1e-15, ...){
+  if(!missing(times)) assertNumeric(times, min.len = 1)
+  if(!missing(n)) assertIntegerish(n, len = 1)
+  assertNumeric(eps, lower = -1, upper = 1)
+
+  UseMethod("pecs", x)
+}
+
+#' @rdname pecs
+#' @export
+pecs.list = function(x, meas = c("graf","logloss"), times, n, eps = 1e-15, task,...){
+  meas = match.arg(meas)
+
+  assert(all(sapply(x, function(y) !is.null(y$model))), "x must be a list of trained survival learners")
```

Comment:
another check necessary to see if `y` has appropriate class? `x` is a list and elements could be anything.

Reply:
I thought about this but this is caught automatically by `y$model`. Unless the user wants to deliberately trick the function by creating a list containing elements that all have a slot `$model`, I think this is sufficient